src/Analysis/Ast/Impl/Analyzer/Evaluation/ExpressionEval.Collections.cs

```diff
@@ -101,40 +102,47 @@ internal sealed partial class ExpressionEval {
-                        return PythonCollectionType.CreateList(Module, new[] { v1 });
-                    case SetComprehension sc:
-                        var v2 = GetValueFromExpression(sc.Item) ?? UnknownType;
-                        return PythonCollectionType.CreateSet(Module, new[] { v2 });
-                    case DictionaryComprehension dc:
-                        var k = GetValueFromExpression(dc.Key) ?? UnknownType;
-                        var v = GetValueFromExpression(dc.Value) ?? UnknownType;
-                        return new PythonDictionary(new PythonDictionaryType(Interpreter.ModuleResolution.BuiltinsModule), new Dictionary<IMember, IMember> { { k, v } });
+        public IMember GetValueFromComprehensionInScope(Comprehension node, bool is3X = true) {
+            if (is3X) {
+                using (OpenScope(Module, node)) {
+                    ProcessComprehension(node);
+                    return GetValueFromComprehension(node);
                 }
+            }
 
-                return UnknownType;
-            } finally {
-                // Remove temporary variables since this is assignment and the right hand
-                // side comprehension does not leak internal variables into the scope.
-                var newVariables = CurrentScope.Variables.ToDictionary(k => k.Name, v => v);
-                var variables = (VariableCollection)CurrentScope.Variables;
-                foreach (var kvp in newVariables) {
-                    if (!oldVariables.ContainsKey(kvp.Key)) {
-                        variables.RemoveVariable(kvp.Key);
-                    }
+            ProcessComprehension(node);
+            return GetValueFromComprehension(node);
+        }
+
+        private IMember GetValueFromComprehension(Comprehension node) {
+            // TODO: Evaluate comprehensions to produce exact contents, if possible.
+            switch (node) {
+                case ListComprehension lc:
+                    var v1 = GetValueFromExpression(lc.Item) ?? UnknownType;
+                    return PythonCollectionType.CreateList(Module, new[] { v1 });
+                case SetComprehension sc:
+                    var v2 = GetValueFromExpression(sc.Item) ?? UnknownType;
+                    return PythonCollectionType.CreateSet(Module, new[] { v2 });
+                case DictionaryComprehension dc:
+                    var k = GetValueFromExpression(dc.Key) ?? UnknownType;
+                    var v = GetValueFromExpression(dc.Value) ?? UnknownType;
+                    return new PythonDictionary(new PythonDictionaryType(Interpreter.ModuleResolution.BuiltinsModule), new Dictionary<IMember, IMember> { { k, v } });
+            }
+            return UnknownType;
+        }
+
+        internal void ProcessComprehensionInScope(Comprehension node, bool is3X = true) {
+            // In 2X there is a bug where comprehension variables get leaked to outer scopes, fixed in 3X
```

Comment:
Nit: I don't think it was a bug, but just how comprehension scoping worked in 2x.

Reply:
OK